src/precompile.c

```diff
@@ -331,16 +331,16 @@ static void jl_compile_all_defs(void)
 static int precompile_enq_specialization_(jl_method_instance_t *mi, void *closure)
 {
     assert(jl_is_method_instance(mi));
-    jl_code_instance_t *codeinst = mi->cache;
+    jl_code_instance_t *codeinst = jl_atomic_load_relaxed(&mi->cache);
     while (codeinst) {
         int do_compile = 0;
-        if (codeinst->invoke != jl_fptr_const_return) {
+        if (jl_atomic_load_relaxed(&codeinst->invoke) != jl_fptr_const_return) {
             if (codeinst->inferred && codeinst->inferred != jl_nothing &&
                 jl_ir_flag_inferred((jl_array_t*)codeinst->inferred) &&
                 !jl_ir_flag_inlineable((jl_array_t*)codeinst->inferred)) {
                 do_compile = 1;
             }
-            else if (codeinst->invoke != NULL || codeinst->precompile) {
+            else if (jl_atomic_load_relaxed(&codeinst->invoke) != NULL || jl_atomic_load_relaxed(&codeinst->precompile)) {
```

Comment:
Maybe fuse these loads into a single one as well?

Reply:
we don't expect this to run with threads, but we likely permit the compiler to fuse these anyways (the result should be the same either way, based on how we define and use this `invoke` field)